### DISCUSSION TITLE
How to download on client?

### DISCUSSION THREAD

**Discussion by mstrop** [exceljs/exceljs#2576] [Q&A]
Is there a way, how to create a XLSX on a client and download it (send it as download file) without saving it or interacting with a server?

**Comment by kirtiesl** [comment]
Hello, looking for the same feature. Please update

**Comment by gtskaushik** [comment]
Yes it is possible.

1. Create the workbook and populate it with data

```js
  res.attachment("test.xlsx")
  workbook.xlsx.write(res)
                .then(function() {
                    res.end()
            });
```

Here res is the response to the client

**Comment by pleger** [comment]
Hi, everyone,
     I am a newest in this topic. Thereby, I couldn't understand the following code in the example:
        `res.attachment("test.xlsx");`
     What is `res` on the client? How can I create `res`?

Thanks, Paul

**Comment by Amerzel** [comment]
I think the previous commenter might have been confused about what was being asked. ```res``` indicates to me that they meant a node response sent back to the client. The original poster was asking for a way to create and download a file without using a server. I'd like this functionality as well and haven't been able to get it to work yet.

**Comment by rojasjandro89** [comment]
Same here, I'm using angular, I have all I need in the client, it would be great if I could just download the file right from the client. So far all I've found has node's "res"

**Comment by pleger** [comment]
Hi, 
  The following code works for me:

```js
workbook.xlsx.writeBuffer().then(function (data) {
    var blob = new Blob([data], {type: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"});
    saveAs(blob, “fileName.xlsx");
});
```

“saveAs” … I used “FileSaver.js”.

Best,

 <http://pleger.cl/>

**Comment by rojasjandro89** [comment]
It works for me too. Thanks a lot!

**Comment by Mafi78** [comment]
Hi,
I've attached one working zip file in this incident: https://github.com/guyonroche/exceljs/issues/322 perhaby cou can try that ?
Best regards
Manfred

**Comment by crib8** [comment]
@pleger do you example on how to download excel? i already used res.download. but still cant download the file. But it can do write excel

**Comment by pleger** [comment]
I hope the following code can help you!

extract from http://pleger.cl/gp

```
$(document).ready(function () {  
  | var workbook = new ExcelJS.Workbook();
  |  
  | workbook.creator = 'Paul Leger';
  | workbook.lastModifiedBy = 'Paul Leger';
  | workbook.created = new Date();
  | workbook.modified = new Date();
  | workbook.lastPrinted = new Date();
  |  
  | var worksheet = workbook.addWorksheet("Publications");
  | worksheet.views = [
  | {state: 'frozen', xSplit: 0, ySplit: 1}
  | ];
  |  
  | worksheet.autoFilter = {
  | from: 'A1',
  | to: 'M1'
  | };
  |  
  | worksheet.columns = [
  | { header: 'Index', key: 'Index', width: 15 },
  | { header: 'Title', key: 'title', width: 25, style: {alignment: {wrapText: true} } },
  | { header: 'Authors', key: 'authors', width: 20, style: {alignment: {wrapText: true} } },
  | { header: 'Journal/Conference', key: 'jc', width: 25, style: {alignment: {wrapText: true} } },
  | { header: 'Type', key: 'type', width: 12, style: {alignment: {wrapText: true} } },
  | { header: 'Year', key: 'year', width: 12, style: {numFmt: "0000"}},
  | { header: 'Month', key: 'month', width: 12},
  | { header: 'volume', key: 'volume', width: 12},
  | { header: 'number', key: 'number', width: 12},
  | { header: 'Pages', key: 'pages', width: 12},
  | { header: 'Location', key: 'location', width: 20, style: {alignment: {wrapText: true} } },
  | { header: 'doi', key: 'doi', width: 22, style: {alignment: {wrapText: true} } },
  | { header: 'affiliation', key: 'affiliation', width: 20, style: {alignment: {wrapText: true} } }
  | ];
  |  
  | var firstRow = worksheet.getRow(1);
  | firstRow.font = { name: 'New Times Roman', family: 4, size: 10, bold: true, color: {argb:'80EF1C1C'} };
  | firstRow.alignment = { vertical: 'middle', horizontal: 'center'};
  | firstRow.height = 20;
  |  
  |  
  | csv.shift();
  | worksheet.addRows(csv);
  |  
  | var buff = workbook.xlsx.writeBuffer().then(function (data) {
  | var blob = new Blob([data], {type: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"});
  | saveAs(blob, "publications.xlsx");
  | });

```

**Comment by crib8** [comment]
Thanks you so much!

**Comment by LiHaoGit** [comment]
It works for me

`koa2`
```
export async function download(ctx: koa.Context) {
    const workbook = new Excel.Workbook()
    const worksheet = workbook.addWorksheet("report")
    worksheet.columns = [
      {header: "Date", key: "date", width: 20}]
    ctx.response.attachment("report.xlsx")
    ctx.status = 200
    await workbook.xlsx.write(ctx.res)
    ctx.res.end()
}
```

**Comment by Messilimeng** [comment]
@LiHaoGit 好啊 我正在用这个插件 可是一直下载不了 能不能加个qq 咨询一下

**Comment by shikhaBasra** [comment]
`xlsx.writeBuffer()`
this code gives me error that function not exist

**Comment by dlinx** [comment]
To download on browser.

 ```
workbook.xlsx.writeBuffer().then(function (data: Blob) {
      const blob = new Blob([data],
        { type: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet' });
      const url = window.URL.createObjectURL(blob);
      const anchor = document.createElement('a');
      anchor.href = url;
      anchor.download = 'download.xls';
      anchor.click();
      window.URL.revokeObjectURL(url);
    });
```

**Comment by JarvisQJ** [comment]
> It works for me
> 
> `koa2`
> 
> ```
> export async function download(ctx: koa.Context) {
>     const workbook = new Excel.Workbook()
>     const worksheet = workbook.addWorksheet("report")
>     worksheet.columns = [
>       {header: "Date", key: "date", width: 20}]
>     ctx.response.attachment("report.xlsx")
>     ctx.status = 200
>     await workbook.xlsx.write(ctx.res)
>     ctx.res.end()
> }
> ```

@LiHaoGit hi,how to receive and turn the data to excel file when you send data like this?

**Comment by Siemienik** [comment]
#269 #37

**Comment by praveendev404** [comment]
res.status(200);
res.setHeader('Content-Type', 'text/xlsx');
res.setHeader(
'Content-Disposition',
'attachment; filename=test.xlsx'
);
workbook.xlsx.write(res)
.then(function () {
res.end()
});

**Comment by chris-canipe** [comment]
> To download on browser.
> 
> ```
> workbook.xlsx.writeBuffer().then(function (data: Blob) {
>      const blob = new Blob([data],
>        { type: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet' });
>      const url = window.URL.createObjectURL(blob);
>      const anchor = document.createElement('a');
>      anchor.href = url;
>      anchor.download = 'download.xls';
>      anchor.click();
>      window.URL.revokeObjectURL(url);
>    });
> ```

I had to change `anchor.click();` to `anchor.dispatchEvent(new MouseEvent('click'));` to get this working in Firefox. Thanks!

**Comment by mateo2181** [comment]
Someone could do this sending the buffer from the backend (Node) ?

**Comment by MeghaShrivastava** [comment]
> I hope the following code can help you!
> 
> extract from http://pleger.cl/gp
> 
> ```
> $(document).ready(function () {  
>   | var workbook = new ExcelJS.Workbook();
>   |  
>   | workbook.creator = 'Paul Leger';
>   | workbook.lastModifiedBy = 'Paul Leger';
>   | workbook.created = new Date();
>   | workbook.modified = new Date();
>   | workbook.lastPrinted = new Date();
>   |  
>   | var worksheet = workbook.addWorksheet("Publications");
>   | worksheet.views = [
>   | {state: 'frozen', xSplit: 0, ySplit: 1}
>   | ];
>   |  
>   | worksheet.autoFilter = {
>   | from: 'A1',
>   | to: 'M1'
>   | };
>   |  
>   | worksheet.columns = [
>   | { header: 'Index', key: 'Index', width: 15 },
>   | { header: 'Title', key: 'title', width: 25, style: {alignment: {wrapText: true} } },
>   | { header: 'Authors', key: 'authors', width: 20, style: {alignment: {wrapText: true} } },
>   | { header: 'Journal/Conference', key: 'jc', width: 25, style: {alignment: {wrapText: true} } },
>   | { header: 'Type', key: 'type', width: 12, style: {alignment: {wrapText: true} } },
>   | { header: 'Year', key: 'year', width: 12, style: {numFmt: "0000"}},
>   | { header: 'Month', key: 'month', width: 12},
>   | { header: 'volume', key: 'volume', width: 12},
>   | { header: 'number', key: 'number', width: 12},
>   | { header: 'Pages', key: 'pages', width: 12},
>   | { header: 'Location', key: 'location', width: 20, style: {alignment: {wrapText: true} } },
>   | { header: 'doi', key: 'doi', width: 22, style: {alignment: {wrapText: true} } },
>   | { header: 'affiliation', key: 'affiliation', width: 20, style: {alignment: {wrapText: true} } }
>   | ];
>   |  
>   | var firstRow = worksheet.getRow(1);
>   | firstRow.font = { name: 'New Times Roman', family: 4, size: 10, bold: true, color: {argb:'80EF1C1C'} };
>   | firstRow.alignment = { vertical: 'middle', horizontal: 'center'};
>   | firstRow.height = 20;
>   |  
>   |  
>   | csv.shift();
>   | worksheet.addRows(csv);
>   |  
>   | var buff = workbook.xlsx.writeBuffer().then(function (data) {
>   | var blob = new Blob([data], {type: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"});
>   | saveAs(blob, "publications.xlsx");
>   | });
> ```

Hello,
My data is coming from API but after downloading the excel the data is not having proper sequence like the data which is in column 1 of my report is now coming in column 4 and so on.

I'm using calling data list and header like this:
generateExcel(){
    this._http.generateExcel(this.headerInfo, this.spendclassificationlist);
  }

Could you please suggest me how can I map my data with the proper header.

**Comment by huy-lv** [comment]
> It works for me
> 
> `koa2`
> 
> ```
> export async function download(ctx: koa.Context) {
>     const workbook = new Excel.Workbook()
>     const worksheet = workbook.addWorksheet("report")
>     worksheet.columns = [
>       {header: "Date", key: "date", width: 20}]
>     ctx.response.attachment("report.xlsx")
>     ctx.status = 200
>     await workbook.xlsx.write(ctx.res)
>     ctx.res.end()
> }
> ```

What about client side implementation?

**Comment by Siemienik** [comment]
The easiest way is by using filesaver: https://www.npmjs.com/package/file-saver

**Comment by luopeihai** [comment]
The font end needs the responseType blob：
```js
  axios({
      method: "get",
      url: "http://localhost:3000/export/xls",
      responseType: "blob", // 重要：指定响应类型为blob
    })
      .then((response) => {
        const url = window.URL.createObjectURL(
          new Blob([response.data])
        );
        const link = document.createElement("a");
        link.href = url;
        link.setAttribute("download", "data.xls"); // 设置下载文件名
        document.body.appendChild(link);
        link.click();
        document.body.removeChild(link);
        window.URL.revokeObjectURL(url);
      })
      .catch((error) => {
        console.error("导出失败:", error);
      });
```

The server requires a response buffer:
```node
const workbook = new ExcelJS.Workbook();
        const worksheet = workbook.addWorksheet('TestExportXLS');

        worksheet.columns = [
            { header: 'name', key: 'name' },
            { header: 'age', key: 'age' }
        ];

        worksheet.addRow({
            name: 'Neo Luo',
            age: 18
        });

        const buffer = await workbook.xlsx.writeBuffer();
        res.header('Content-Disposition', 'attachment; filename=anlikodullendirme.xlsx');
        res.type('application/vnd.openxmlformats-officedocument.spreadsheetml.sheet');
        res.send(buffer);
```
[detail](https://medium.com/@ggluopeihai/nestjs-export-excel-file-697e3891ea8f)

**Comment by xoewrl** [comment]
vue demo

``` js 
<script setup>
import * as ExcelJS from "exceljs";

const download = () => {
  let workbook = new ExcelJS.Workbook();
  workbook.creator = 'Paul Leger';
  workbook.lastModifiedBy = 'Paul Leger';
  workbook.created = new Date();
  workbook.modified = new Date();
  workbook.lastPrinted = new Date();
  let worksheet = workbook.addWorksheet("Publications");
  worksheet.views = [
    {state: 'frozen', xSplit: 0, ySplit: 1}
  ];
  worksheet.autoFilter = {
    from: 'A1',
    to: 'M1'
  };
  worksheet.columns = [
    {header: 'Index', key: 'Index', width: 15},
    {header: 'Title', key: 'title', width: 25, style: {alignment: {wrapText: true}}},
    {header: 'Authors', key: 'authors', width: 20, style: {alignment: {wrapText: true}}},
    {header: 'Journal/Conference', key: 'jc', width: 25, style: {alignment: {wrapText: true}}},
    {header: 'Type', key: 'type', width: 12, style: {alignment: {wrapText: true}}},
    {header: 'Year', key: 'year', width: 12, style: {numFmt: "0000"}},
    {header: 'Month', key: 'month', width: 12},
    {header: 'volume', key: 'volume', width: 12},
    {header: 'number', key: 'number', width: 12},
    {header: 'Pages', key: 'pages', width: 12},
    {header: 'Location', key: 'location', width: 20, style: {alignment: {wrapText: true}}},
    {header: 'doi', key: 'doi', width: 22, style: {alignment: {wrapText: true}}},
    {header: 'affiliation', key: 'affiliation', width: 20, style: {alignment: {wrapText: true}}}
  ];

  let firstRow = worksheet.getRow(1);
  firstRow.font = {name: 'New Times Roman', family: 4, size: 10, bold: true, color: {argb: '80EF1C1C'}};
  firstRow.alignment = {vertical: 'middle', horizontal: 'center'};
  firstRow.height = 20;

  // csv.shift();
  // worksheet.addRows(csv);
  workbook.xlsx.writeBuffer().then(function (data) {
    const blob = new Blob([data],
        // { type: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet' }
        {type: 'application/vnd.ms-excel'}
    );
    const url = window.URL.createObjectURL(blob);
    const anchor = document.createElement('a');
    anchor.href = url;
    anchor.download = 'download.xlsx';
    anchor.click();
    window.URL.revokeObjectURL(url);
  });
}

</script>
<template>
  <el-button style="margin: 50px" @click="download"> 下载</el-button>
</template>

<style scoped>

</style>
```

**Comment by xoewrl** [comment]
demo2
```js 
<script setup>
import * as ExcelJS from "exceljs";

const download = () => {
  let workbook = new ExcelJS.Workbook();
  workbook.creator = 'auto';
  workbook.lastModifiedBy = 'auto';
  workbook.created = new Date();
  workbook.modified = new Date();
  workbook.lastPrinted = new Date();
  let worksheet = workbook.addWorksheet("Sheet1");
  worksheet.columns = [
    {header: '序号', key: 'c1', width: 20},
    {header: '名称', key: 'c2', width: 20,  style: {numFmt: "@"}},
    {header: '日期', key: 'c3', width: 12, style: {numFmt: "@"}},  // @ stands for text
  ];

  let firstRow = worksheet.getRow(1);
  firstRow.font = {name: 'New Times Roman', family: 4, size: 10, bold: true};
  firstRow.alignment = {vertical: 'middle', horizontal: 'center'};
  firstRow.height = 20;

  worksheet.addRows(
      [
          {c1: '1', c2: 2, c3: '2025/06'},
          {c1: '2', c2: '2', c3: '2025/01'},
      ]
  )

  workbook.xlsx.writeBuffer().then(function (data) {
    const blob = new Blob([data],
        {type: 'application/vnd.ms-excel'}
    );
    const url = window.URL.createObjectURL(blob);
    const anchor = document.createElement('a');
    anchor.href = url;
    anchor.download = 'download.xlsx';
    anchor.click();
    window.URL.revokeObjectURL(url);
  });
}

</script>
<template>
  <el-button style="margin: 50px" @click="download"> 下载</el-button>
</template>

<style scoped>

</style>
```
